File: backend/app/routes/resume.py
"""
Resume API routes for managing user resumes.

Provides CRUD operations for resumes stored in Firebase.
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..dependencies.rate_limiting import rate_limited_core_user
from ..schemas.auth import UserToken
from ..schemas.resume import (
    ResumeData,
    ResumeSummary,
    CreateResumeRequest,
    UpdateResumeRequest,
    ResumeListResponse,
)
from ..services.resume_service import (
    get_resume_service,
    ResumeServiceError,
    ResumeNotFoundError,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/resumes", tags=["resumes"])


class CreateResumeResponse(BaseModel):
    """Response for resume creation."""

    id: str
    message: str


class DuplicateResumeRequest(BaseModel):
    """Request for duplicating a resume."""

    new_name: Optional[str] = None


class DuplicateResumeResponse(BaseModel):
    """Response for resume duplication."""

    id: str
    message: str


class DeleteResumeResponse(BaseModel):
    """Response for resume deletion."""

    success: bool
    message: str


@router.post("/", response_model=CreateResumeResponse)
def create_resume(
    request: CreateResumeRequest,
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    Create a new resume.

    Creates a new resume with a unique ID and stores it in the user's
    resume collection.
    """
    try:
        resume_service = get_resume_service()
        resume_id = resume_service.create_resume(user.uid, request)
        return CreateResumeResponse(id=resume_id, message="Resume created successfully")
    except ResumeServiceError as e:
        logger.error(f"Failed to create resume for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to create resume")
    except Exception as e:
        logger.error(f"Unexpected error creating resume for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/", response_model=ResumeListResponse)
def list_resumes(
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    List all resumes for the current user.

    Returns a list of resume summaries with names and timestamps.
    """
    try:
        resume_service = get_resume_service()
        summaries = resume_service.list_resumes(user.uid)
        return ResumeListResponse(resumes=summaries, total=len(summaries))
    except ResumeServiceError as e:
        logger.error(f"Failed to list resumes for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to list resumes")
    except Exception as e:
        logger.error(f"Unexpected error listing resumes for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/{resume_id}", response_model=ResumeData)
def get_resume(
    resume_id: str,
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    Get a single resume by ID.

    Returns the full resume data for the specified resume.
    """
    try:
        resume_service = get_resume_service()
        resume = resume_service.get_resume(user.uid, resume_id)
        return resume
    except ResumeNotFoundError:
        raise HTTPException(status_code=404, detail="Resume not found")
    except ResumeServiceError as e:
        logger.error(f"Failed to get resume {resume_id} for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve resume")
    except Exception as e:
        logger.error(
            f"Unexpected error getting resume {resume_id} for user {user.uid}: {e}"
        )
        raise HTTPException(status_code=500, detail="Internal server error")


@router.put("/{resume_id}", response_model=dict)
def update_resume(
    resume_id: str,
    request: UpdateResumeRequest,
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    Update an existing resume.

    Updates the specified fields of the resume. Only provided fields
    will be updated; others remain unchanged.
    """
    try:
        resume_service = get_resume_service()
        resume_service.update_resume(user.uid, resume_id, request)
        return {"message": "Resume updated successfully"}
    except ResumeNotFoundError:
        raise HTTPException(status_code=404, detail="Resume not found")
    except ResumeServiceError as e:
        logger.error(f"Failed to update resume {resume_id} for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to update resume")
    except Exception as e:
        logger.error(
            f"Unexpected error updating resume {resume_id} for user {user.uid}: {e}"
        )
        raise HTTPException(status_code=500, detail="Internal server error")


@router.delete("/{resume_id}", response_model=DeleteResumeResponse)
def delete_resume(
    resume_id: str,
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    Delete a resume.

    Permanently removes the resume from the user's collection.
    """
    try:
        resume_service = get_resume_service()
        resume_service.delete_resume(user.uid, resume_id)
        return DeleteResumeResponse(success=True, message="Resume deleted successfully")
    except ResumeNotFoundError:
        raise HTTPException(status_code=404, detail="Resume not found")
    except ResumeServiceError as e:
        logger.error(f"Failed to delete resume {resume_id} for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to delete resume")
    except Exception as e:
        logger.error(
            f"Unexpected error deleting resume {resume_id} for user {user.uid}: {e}"
        )
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/{resume_id}/duplicate", response_model=DuplicateResumeResponse)
def duplicate_resume(
    resume_id: str,
    request: DuplicateResumeRequest = None,
    user: UserToken = Depends(rate_limited_core_user),
):
    """
    Duplicate an existing resume.

    Creates a copy of the resume with a new ID. Optionally provide
    a new name for the copy.
    """
    try:
        resume_service = get_resume_service()
        new_name = request.new_name if request else None
        new_id = resume_service.duplicate_resume(user.uid, resume_id, new_name)
        return DuplicateResumeResponse(
            id=new_id, message="Resume duplicated successfully"
        )
    except ResumeNotFoundError:
        raise HTTPException(status_code=404, detail="Resume not found")
    except ResumeServiceError as e:
        logger.error(f"Failed to duplicate resume {resume_id} for user {user.uid}: {e}")
        raise HTTPException(status_code=500, detail="Failed to duplicate resume")
    except Exception as e:
        logger.error(
            f"Unexpected error duplicating resume {resume_id} for user {user.uid}: {e}"
        )
        raise HTTPException(status_code=500, detail="Internal server error")

File: apps/main/src/lib/api/resume.ts
/**
 * Resume API service for communicating with the backend
 *
 * Provides CRUD operations for resumes stored in Firebase.
 * Follows the same patterns as portfolio.ts for consistency.
 */

import type {
  ResumeData,
  ResumeSummary,
  CreateResumeRequest,
  UpdateResumeRequest,
} from "@/types/resume";
import { getIdToken } from "../firebase";
import { env } from "@/lib/env";

const API_BASE_URL = env.API_BASE_URL;

export class ResumeAPIError extends Error {
  constructor(message: string, public status?: number, public code?: string) {
    super(message);
    this.name = "ResumeAPIError";
  }
}

/**
 * Get authorization headers with Firebase ID token
 */
async function getAuthHeaders(): Promise<HeadersInit> {
  const token = await getIdToken();
  if (!token) {
    throw new ResumeAPIError("User not authenticated", 401, "AUTH_REQUIRED");
  }

  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };
}

/**
 * Handle API response and errors
 */
async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    let errorMessage = `HTTP ${response.status}`;
    let errorCode = "API_ERROR";

    try {
      const errorData = await response.json();
      // Handle cases where detail is an object (validation errors)
      if (typeof errorData.detail === "object") {
        errorMessage = JSON.stringify(errorData.detail);
      } else {
        errorMessage = errorData.detail || errorMessage;
      }
      errorCode = errorData.error_code || errorCode;
      console.error("API Error Details:", errorData);
    } catch {
      errorMessage = response.statusText || errorMessage;
    }

    throw new ResumeAPIError(errorMessage, response.status, errorCode);
  }

  if (
    response.status === 204 ||
    response.headers.get("content-length") === "0"
  ) {
    return null as T;
  }

  try {
    return await response.json();
  } catch {
    throw new ResumeAPIError("Invalid response format", response.status);
  }
}

/** Response from list resumes endpoint */
interface ResumeListResponse {
  resumes: ResumeSummary[];
  total: number;
}

/** Response from create resume endpoint */
interface CreateResumeResponse {
  id: string;
  message: string;
}

/** Response from duplicate resume endpoint */
interface DuplicateResumeResponse {
  id: string;
  message: string;
}

/** Response from delete resume endpoint */
interface DeleteResumeResponse {
  success: boolean;
  message: string;
}

/**
 * Ensure resume data has all required fields with defaults
 */
function normalizeResumeData(data: Record<string, unknown>): ResumeData {
  const result = { ...data } as ResumeData;

  // Ensure achievements exists
  result.achievements = (data.achievements as string[]) || [];

  // Ensure profiles exists in personal_info
  if (result.personal_info) {
    result.personal_info = {
      ...result.personal_info,
      profiles: result.personal_info.profiles || {
        linkedin: null,
        github: null,
        leetcode: null,
        codeforces: null,
        codechef: null,
        website: null,
      },
    };
  }

  // Ensure section_order includes achievements
  if (result.section_order && !result.section_order.includes("achievements")) {
    result.section_order = [...result.section_order, "achievements"];
  }

  return result;
}

/**
 * Create a new resume
 * @param data - Resume data to create
 * @returns The ID of the created resume
 */
export async function createResume(data: CreateResumeRequest): Promise<string> {
  try {
    const headers = await getAuthHeaders();

    const response = await fetch(`${API_BASE_URL}/api/resumes/`, {
      method: "POST",
      headers,
      body: JSON.stringify(data),
    });

    const result = await handleResponse<CreateResumeResponse>(response);
    return result.id;
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error creating resume:", error);
    throw new ResumeAPIError(
      "Failed to create resume",
      undefined,
      "CREATE_ERROR"
    );
  }
}

/**
 * Get a single resume by ID
 * @param resumeId - The ID of the resume to retrieve
 * @returns The full resume data
 */
export async function getResume(resumeId: string): Promise<ResumeData> {
  try {
    const headers = await getAuthHeaders();

    const response = await fetch(`${API_BASE_URL}/api/resumes/${resumeId}`, {
      method: "GET",
      headers,
    });

    const data = await handleResponse<Record<string, unknown>>(response);
    return normalizeResumeData(data);
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error fetching resume:", error);
    throw new ResumeAPIError(
      "Failed to fetch resume",
      undefined,
      "FETCH_ERROR"
    );
  }
}

/**
 * List all resumes for the current user
 * @returns List of resume summaries with total count
 */
export async function listResumes(): Promise<ResumeListResponse> {
  try {
    const headers = await getAuthHeaders();

    const response = await fetch(`${API_BASE_URL}/api/resumes/`, {
      method: "GET",
      headers,
    });

    return await handleResponse<ResumeListResponse>(response);
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error listing resumes:", error);
    throw new ResumeAPIError("Failed to list resumes", undefined, "LIST_ERROR");
  }
}

/**
 * Update an existing resume
 * @param resumeId - The ID of the resume to update
 * @param data - Partial resume data to update
 */
export async function updateResume(
  resumeId: string,
  data: UpdateResumeRequest
): Promise<void> {
  try {
    const headers = await getAuthHeaders();

    const response = await fetch(`${API_BASE_URL}/api/resumes/${resumeId}`, {
      method: "PUT",
      headers,
      body: JSON.stringify(data),
    });

    await handleResponse<{ message: string }>(response);
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error updating resume:", error);
    throw new ResumeAPIError(
      "Failed to update resume",
      undefined,
      "UPDATE_ERROR"
    );
  }
}

/**
 * Delete a resume
 * @param resumeId - The ID of the resume to delete
 */
export async function deleteResume(resumeId: string): Promise<void> {
  try {
    const headers = await getAuthHeaders();

    const response = await fetch(`${API_BASE_URL}/api/resumes/${resumeId}`, {
      method: "DELETE",
      headers,
    });

    await handleResponse<DeleteResumeResponse>(response);
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error deleting resume:", error);
    throw new ResumeAPIError(
      "Failed to delete resume",
      undefined,
      "DELETE_ERROR"
    );
  }
}

/**
 * Duplicate an existing resume
 * @param resumeId - The ID of the resume to duplicate
 * @param newName - Optional new name for the duplicated resume
 * @returns The ID of the new resume
 */
export async function duplicateResume(
  resumeId: string,
  newName?: string
): Promise<string> {
  try {
    const headers = await getAuthHeaders();

    const body = newName ? { new_name: newName } : {};

    const response = await fetch(
      `${API_BASE_URL}/api/resumes/${resumeId}/duplicate`,
      {
        method: "POST",
        headers,
        body: JSON.stringify(body),
      }
    );

    const result = await handleResponse<DuplicateResumeResponse>(response);
    return result.id;
  } catch (error) {
    if (error instanceof ResumeAPIError) {
      throw error;
    }

    console.error("Error duplicating resume:", error);
    throw new ResumeAPIError(
      "Failed to duplicate resume",
      undefined,
      "DUPLICATE_ERROR"
    );
  }
}
